Jenkinsfile

```diff
@@ -154,13 +162,50 @@ node {
       if (params.BUILD_GRAVITY_APP) {
         withEnv(MAKE_ENV) {
           sh 'make export'
-          archiveArtifacts "build/application.tar"
         }
       } else {
         echo 'skipped build gravity app'
       }
     }
+
+    stage('upload application tarball to S3') {
+      if (isProtectedBranch(env.TAG) && params.IMPORT_APP_PACKAGE) {
+        withCredentials([usernamePassword(credentialsId: "${AWS_CREDENTIALS}", usernameVariable: 'AWS_ACCESS_KEY_ID', passwordVariable: 'AWS_SECRET_ACCESS_KEY')]) {
+          def artifactName = "application.tar"
+          def s3AppName = "stolon"
+          if (params.BUILD_GRAVITY_HELM_APP) {
+            artifactName = "helm-application.tar"
+            s3AppName = "stolon-helm"
+          }
+          def S3_URL = "s3://${S3_UPLOAD_PATH}/${s3AppName}:${APP_VERSION}.tar"
+          withEnv(MAKE_ENV + ["S3_URL=${S3_URL}"]) {
+            sh 'aws s3 cp --only-show-errors build/${artifactName} ${S3_URL}'
+          }
+        }
+      }
+    }
+  }
+}
+
+def isProtectedBranch(branchOrTagName) {
+  // check if tag or branch empty
+  if (!branchOrTagName?.trim()) {
+    return false
+  }
+
+  String[] protectedBranches = ['master', 'support/1.12.x']
```

Comment:
the pattern for support branches can be `support/*` as all support branches are protected.